File: DynamicKey/AgoraDynamicKey/php/test/ApaasTokenBuilderTest.php
<?php

require_once "../src/AccessToken2.php";
require_once "../src/ApaasTokenBuilder.php";

class ApaasTokenBuilderTest
{
    public $appId = "970CA35de60c44645bbae8a215061b33";
    public $appCertificate = "5CFd2fd1755d40ecb72977518be15d3b";
    public $expire = 600;
    public $roomUuid = "123";
    public $userUuid = "2882341273";
    public $role = 1;

    /**
     * Run all APaaS token builder test cases.
     */
    public function run()
    {
        $this->test_buildRoomUserToken();
        $this->test_buildUserToken();
        $this->test_buildAppToken();
    }

    /**
     * Verify APaaS room user token generation and parsing.
     */
    public function test_buildRoomUserToken()
    {
        $token = ApaasTokenBuilder::buildRoomUserToken($this->appId, $this->appCertificate, $this->roomUuid, $this->userUuid, $this->role, $this->expire);
        $accessToken = new AccessToken2();
        $accessToken->parse($token);
        $serviceApaas = $accessToken->getServices(ServiceApaas::SERVICE_TYPE)[0];

        Util::assertEqual($this->appId, $accessToken->appId);
        Util::assertEqual($this->expire, $accessToken->expire);
        Util::assertEqual($this->roomUuid, $serviceApaas->roomUuid);
        Util::assertEqual($this->userUuid, $serviceApaas->userUuid);
        Util::assertEqual($this->role, $serviceApaas->role);
        Util::assertEqual($this->expire, $serviceApaas->privileges[ServiceApaas::PRIVILEGE_ROOM_USER]);
    }

    /**
     * Verify APaaS user token generation and parsing.
     */
    public function test_buildUserToken()
    {
        $token = ApaasTokenBuilder::buildUserToken($this->appId, $this->appCertificate, $this->userUuid, $this->expire);
        $accessToken = new AccessToken2();
        $accessToken->parse($token);
        $serviceApaas = $accessToken->getServices(ServiceApaas::SERVICE_TYPE)[0];

        Util::assertEqual($this->appId, $accessToken->appId);
        Util::assertEqual($this->expire, $accessToken->expire);
        Util::assertEqual($this->userUuid, $serviceApaas->userUuid);
        Util::assertEqual("", $serviceApaas->roomUuid);
        Util::assertEqual(-1, $serviceApaas->role);
        Util::assertEqual($this->expire, $serviceApaas->privileges[ServiceApaas::PRIVILEGE_USER]);
    }

    /**
     * Verify APaaS application token generation and parsing.
     */
    public function test_buildAppToken()
    {
        $token = ApaasTokenBuilder::buildAppToken($this->appId, $this->appCertificate, $this->expire);
        $accessToken = new AccessToken2();
        $accessToken->parse($token);
        $serviceApaas = $accessToken->getServices(ServiceApaas::SERVICE_TYPE)[0];

        Util::assertEqual($this->appId, $accessToken->appId);
        Util::assertEqual($this->expire, $accessToken->expire);
        Util::assertEqual($this->expire, $serviceApaas->privileges[ServiceApaas::PRIVILEGE_APP]);

        Util::assertEqual("", $serviceApaas->roomUuid);
        Util::assertEqual("", $serviceApaas->userUuid);
        Util::assertEqual(-1, $serviceApaas->role);
    }
}

$apaasTokenBuilderTest = new ApaasTokenBuilderTest();
$apaasTokenBuilderTest->run();

File: DynamicKey/AgoraDynamicKey/dart/test/apaas_token_builder_test.dart
import 'package:agora_token_generator/agora_token_generator.dart';
import 'package:flutter_test/flutter_test.dart';

/// Runs APaaS Token007 Builder tests.
void main() {
  const appId = '970CA35de60c44645bbae8a215061b33';
  const appCertificate = '5CFd2fd1755d40ecb72977518be15d3b';
  const roomUuid = '123';
  const userUuid = '2882341273';
  const role = 1;
  const expireSeconds = 600;

  group('APaaS Token Builder', () {
    // Verifies room-user tokens contain APaaS, RTM, and Chat services.
    test('buildRoomUserToken builds all required services', () {
      final token = ApaasTokenBuilder.buildRoomUserToken(
        appId: appId,
        appCertificate: appCertificate,
        roomUuid: roomUuid,
        userUuid: userUuid,
        role: role,
        expireSeconds: expireSeconds,
      );

      final parsed = AccessToken.empty();
      expect(parsed.parse(token), isTrue);
      expect(parsed.verifySignature(appCertificate), isTrue);
      expect(parsed.services.map((service) => service.serviceType), [
        Service.RTM,
        Service.CHAT,
        Service.APAAS,
      ]);
      final apaas = parsed.getServices(Service.APAAS).single as ServiceApaas;
      final rtm = parsed.getServices(Service.RTM).single as ServiceRtm;
      final chat = parsed.getServices(Service.CHAT).single as ServiceChat;
      expect(apaas.roomUuid, roomUuid);
      expect(apaas.userUuid, userUuid);
      expect(apaas.role, role);
      expect(
        apaas.privileges[ServiceApaas.privilegeRoomUser],
        expireSeconds,
      );
      expect(rtm.userId, userUuid);
      expect(chat.userId, '6063383428a36fba3fb6030becf8094e');
    });

    // Verifies APaaS user token fields and privilege.
    test('buildUserToken builds a user token', () {
      final token = ApaasTokenBuilder.buildUserToken(
        appId: appId,
        appCertificate: appCertificate,
        userUuid: userUuid,
        expireSeconds: expireSeconds,
      );

      final parsed = AccessToken.empty();
      expect(parsed.parse(token), isTrue);
      final service = parsed.getServices(Service.APAAS).single as ServiceApaas;
      expect(service.roomUuid, isEmpty);
      expect(service.userUuid, userUuid);
      expect(service.role, -1);
      expect(service.privileges[ServiceApaas.privilegeUser], expireSeconds);
    });

    // Verifies APaaS application token fields and privilege.
    test('buildAppToken builds an application token', () {
      final token = ApaasTokenBuilder.buildAppToken(
        appId: appId,
        appCertificate: appCertificate,
        expireSeconds: expireSeconds,
      );

      final parsed = AccessToken.empty();
      expect(parsed.parse(token), isTrue);
      final service = parsed.getServices(Service.APAAS).single as ServiceApaas;
      expect(service.roomUuid, isEmpty);
      expect(service.userUuid, isEmpty);
      expect(service.role, -1);
      expect(service.privileges[ServiceApaas.privilegeApp], expireSeconds);
    });
  });
}
